src/word_search_generator/__init__.py

```diff
@@ -11,36 +10,96 @@
 __app_name__ = "word-search"
 __version__ = "1.4.0"
 
+
 import json
 from pathlib import Path
-from typing import Optional, Union
+from typing import Any, Iterable, Optional, Union
 
-from word_search_generator import config, export, generate, utils
-from word_search_generator.types import Key, Puzzle
+from . import config, export, generate, utils
+from .types import DirectionSet, Key, Puzzle, Wordlist
 
 
 class WordSearch:
     """This class represents a WordSearch object."""
 
     def __init__(
-        self, words: str, level: Optional[int] = None, size: Optional[int] = None
+        self,
+        words: Optional[str] = None,
+        level: Optional[int | str] = None,
+        size: Optional[int] = None,
+        secret_words: Optional[str] = None,
+        secret_level: Optional[int | str] = None,
     ):
         """Initialize a Word Search puzzle.
 
         Args:
-            words (str): words (str): A string of words separated by spaces, commas,
-            or new lines and limited to 30 word max. Will be trimmed if more.
-            level (Optional[int], optional): Difficulty level. Defaults to None.
+            words (Optional[str], optional): A string of words separated by spaces,
+                commas, or new lines. Will be trimmed if more. Defaults to None.
+            level (Optional[int  |  str], optional): Difficulty level or potential
+                word directions. Defaults to 2.
             size (Optional[int], optional): Puzzle size. Defaults to None.
+            secret_words (Optional[str], optional): A string of words separated by
+                spaces, commas, or new lines. Words will be 'secret' meaning they
+                will not be included in the word list. Defaults to None.
+            secret_level (Optional[int  |  str], optional): Difficulty level or
+                potential word directions for 'secret' words. Defaults to None.
         """
-        self.words = utils.cleanup_input(words)
-        self._key: Key = {}
-        self._level: int = 1
+
+        # setup puzzle
         self._puzzle: Puzzle = []
-        self._size: int = 0
         self._solution: Puzzle = []
-        # generate puzzle
-        self.generate(level, size)
+        self._size: int = size if size else 0
+
+        # setup words
+        # TODO: Check max puzzle words should include both regular and secret words
```

Comment:
@duck57, I wanted your opinion on this too. I already created an issue (#25) but couldn't tag you there. `max_puzzle_words` is set to 100 now but what about this edge case? A user supplies 100 regular words and then 50 secret words.

The way it works now (which may be the best approach) is that the `fill_words()` function in generate.py will try to fill all 100 regular words first, then it will try and fill the 50 secret words. Currently, there is no actual check for how many words get placed into the puzzle, the `max_puzzle_words` check only happens in the cli and when processing the input strings. So you could easily end up with more than 100 words right now is the puzzle size would fit them.

My other thought was to subtract the number of secret words (if any) from 100, so 100-50 = 50 in this case, and then only try inserting 50 of the regular words. This would guarantee the max puzzle words would be 100 at most. It would probably be less because of wordfit exceptions but it would be close.

As I said, this is an extreme edge case, it just happened to pop into my mind when I was testing the updated pdf layout. The attached PDF was a recreation of the about scenario, 100 regular words, and 50 secret words were provided and it handled it fine, I was just curious if you thought a change should be made.

P.S. After writing this whole thing out, I think it's fine as is 🤦‍♂️...

```python
from word_search_generator import WordSearch
from word_search_generator.utils import get_random_words
p = WordSearch(get_random_words(100), secret_words=get_random_words(50))
```

The generator placed 96 regular words and 46 secret words, for a total of 142 played words.

```pycon
> len(p.placed_hidden_words)
96
> len(p.placed_secret_words)
46
> len(p.played_words)
142
```

[test4.pdf](https://github.com/joshbduncan/word-search-generator/files/9841152/test4.pdf)

Reply:
I agree with your "fine as is".  Someone who wants to add 100 words plus 50 secret words either already knows what they want to achieve or is simply testing out the system.